contracts/PangolinExchange.sol

```diff
@@ -0,0 +1,166 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.2;
+
+import "@pangolindex/exchange-contracts/contracts/pangolin-periphery/interfaces/IPangolinRouter.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "./IAssetsExchange.sol";
+import "./SupportedAssets.sol";
+
+/**
+ * @title PangolinExchange
+ * @dev Contract allows user to invest into an ERC20 token
+ * This implementation uses the Pangolin DEX
+ */
+contract PangolinExchange is Ownable, IAssetsExchange {
+  /* ========= STATE VARIABLES ========= */
+  IPangolinRouter pangolinRouter;
+  SupportedAssets supportedAssets;
+
+  /* ========= CONSTRUCTOR ========= */
+
+  constructor (address _pangolinRouter, SupportedAssets _supportedAssets) {
+    pangolinRouter = IPangolinRouter(_pangolinRouter);
+    supportedAssets = _supportedAssets;
+  }
+
+  /* ========= MODIFIERS ========= */
+
+  modifier RefundRemainder {
+    _;
+    if (address(this).balance > 0) {
+      (bool success,) = msg.sender.call{value : address(this).balance}("");
+      require(success, "Refund failed");
+    }
+  }
+
+
+  function transferBack(bytes32 _asset) external override {
+    address tokenAddress = supportedAssets.getAssetAddress(_asset);
+    IERC20 token = IERC20(tokenAddress);
+    token.transfer(msg.sender, token.balanceOf(address(this)));
+  }
+
+
+  /**
+   * Buys selected ERC20 token with AVAX using the Pangolin DEX
+   * Refunds unused AVAX to the msg.sender
+   * @dev _token ERC20 token's address
+   * @dev _exactERC20AmountOut amount of the ERC20 token to be bought
+  **/
+  function buyAsset(bytes32 _token, uint256 _exactERC20AmountOut) payable external override RefundRemainder {
```

Comment:
Hmm, I'm not sure if the funds will be refunded if the external call fails and the modifier won't be executed :/

Let's make some order in here as transferBack and refund are conceptually the same. We can have 2 methods refundEth & refundERC20 and the following flow
1) Let the method return true/false
2) Instead of throwing errors let's return false
3) Any call to pangolin that could throw an exception should be wrapped in (success, failure) = pangolin.call
4) In case of failure we should ALWAYS refund and return false
5) In case of success we should emit an event

Reply:
It is already implemented in this way now that the https://github.com/ava-loan/avaloan/pull/24 was merged